### PR TITLE
Add godeps plugin.

### PR DESCRIPTION
**Pull request by kyrofa**
Snapcraft currently has a go plugin, but go has a number of dependency management systems. One very common one is godeps. This PR resolves LP: [#1595981](https://bugs.launchpad.net/snapcraft/+bug/1595981) by adding a plugin for it.

### CODE REVIEW

File: snapcraft/tests/test_commands_list_plugins.py [hunk 1]
```diff
@@ -23,10 +23,10 @@ class ListPluginsCommandTestCase(tests.TestCase):
...
     # plugin list when wrapper at MAX_CHARACTERS_WRAP
     default_plugin_output = (
-        'ant        catkin  copy  gradle  jdk     kernel  maven  '
-        'nodejs             python2  qmake  scons      \n'
-        'autotools  cmake   go    gulp    kbuild  make    nil    '
-        'plainbox-provider  python3  rust   tar-content\n')
+        'ant        catkin  copy  godeps  gulp  kbuild  make   nil     '
+        'plainbox-provider  python3  rust   tar-content\n'
+        'autotools  cmake   go    gradle  jdk   kernel  maven  nodejs  '
+        'python2            qmake    scons\n')
```
Comment (kyrofa) [inline snapcraft/tests/test_commands_list_plugins.py]: Kill me now. I thought we got rid of these?

Comment (elopio) [inline snapcraft/tests/test_commands_list_plugins.py]: we left just one.

**Comment by elopio** [comment]
:+1: thank you.

File: snapcraft/plugins/godeps.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+'sources' topic for the latter.
+
+Additionally, this plugin uses the following plugin-specific keywords:
+
+    - godeps-file:
+      (string)
+      Path to the godeps dependencies file contained within the source
+      (default: dependencies.tsv)
+
+    - go-importpath:
```
Comment (sergiusens) [inline snapcraft/plugins/godeps.py]: can you please make this required?

File: snapcraft/plugins/godeps.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        self._gopath_bin = os.path.join(self._gopath, 'bin')
+        self._gopath_pkg = os.path.join(self._gopath, 'pkg')
+
+    def pull(self):
+        super().pull()
+
+        if self.options.go_importpath:
+            go_package = self.options.go_importpath
+        else:
+            logger.warning(
```
Comment (sergiusens) [inline snapcraft/plugins/godeps.py]: this should not be needed as gopath would be mandatory.

File: snapcraft/plugins/godeps.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    def _run(self, cmd, **kwargs):
+        env = self._build_environment()
+        return self.run(cmd, cwd=self._gopath_src, env=env, **kwargs)
+
+    def _build_environment(self):
+        env = os.environ.copy()
+        env['GOPATH'] = self._gopath
+        env['PATH'] = '{}:{}'.format(
+            os.path.join(self._gopath, 'bin'), env.get('PATH', ''))
+        linker_flags = ' '.join([
```
Comment (sergiusens) [inline snapcraft/plugins/godeps.py]: can't you retrieve this from `common`, I bet you also want the stage package ones here.

File: snapcraft/plugins/godeps.py (hunk 3 above)
Comment (sergiusens) [inline snapcraft/plugins/godeps.py]: the logic here would be greatly simplified if so ;-)

File: snapcraft/plugins/godeps.py (hunk 2 above)
Comment (kyrofa) [inline snapcraft/plugins/godeps.py]: Sure, done.

File: snapcraft/plugins/godeps.py (hunk 3 above)
Comment (kyrofa) [inline snapcraft/plugins/godeps.py]: Indeed!

File: snapcraft/plugins/godeps.py (hunk 4 above)
Comment (kyrofa) [inline snapcraft/plugins/godeps.py]: Haha, I completely forgot about those functions already! How sad. Definitely should use them here-- done.

File: snapcraft/tests/test_commands_list_plugins.py [hunk 5]
```diff
@@ -47,11 +47,11 @@ def test_list_plugins_small_terminal(self):
...
-            'ant        go      kernel  plainbox-provider  scons      \n'
-            'autotools  gradle  make    python2            tar-content\n'
-            'catkin     gulp    maven   python3          \n'
-            'cmake      jdk     nil     qmake            \n'
-            'copy       kbuild  nodejs  rust             \n')
+            'ant        go      kbuild  nodejs             rust       \n'
+            'autotools  godeps  kernel  plainbox-provider  scons      \n'
+            'catkin     gradle  make    python2            tar-content\n'
+            'cmake      gulp    maven   python3          \n'
+            'copy       jdk     nil     qmake            \n')
```
Comment (kyrofa) [inline snapcraft/tests/test_commands_list_plugins.py]: :stuck_out_tongue:

**Comment by nskaggs** [comment]
+1 from me. Tested and working for building juju. This made my juju snapcraft.yaml file much simpler, and added tags support! Thanks!

**Comment by sergiusens** [comment]
:+1:

**Pull request merged**